sentry-core/src/main/java/io/sentry/core/cache/DiskCache.java

```diff
@@ -0,0 +1,152 @@
+package io.sentry.core.cache;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+import static io.sentry.core.ILogger.logIfNotNull;
+import static io.sentry.core.SentryLevel.DEBUG;
+import static io.sentry.core.SentryLevel.ERROR;
+import static io.sentry.core.SentryLevel.WARNING;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import io.sentry.core.ISerializer;
+import io.sentry.core.SentryEvent;
+import io.sentry.core.SentryLevel;
+import io.sentry.core.SentryOptions;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A simple cache implementation storing the events to a disk, each event in a separater file in a configured directory.
+ */
+public final class DiskCache implements IEventCache {
+  /**
+   * File suffix added to all serialized event files.
+   */
+  public static final String FILE_SUFFIX = ".sentry-event";
+
+  @SuppressWarnings("CharsetObjectCanBeUsed")
+  private static final Charset UTF8 = Charset.forName("UTF-8");
+
+  private final File directory;
+  private final int maxSize;
+  private final ISerializer serializer;
+  private final SentryOptions options;
+
+  public DiskCache(SentryOptions options) {
+    this.directory = new File(options.getCacheDirPath());
+    this.maxSize = options.getCacheDirSize();
+    this.serializer = options.getSerializer();
+    this.options = options;
+
+    try {
+      //noinspection ResultOfMethodCallIgnored
+      directory.mkdirs();
+      checkDirectoryValid();
+    } catch (IllegalStateException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new IllegalStateException("Failed to initialize the directory for caching Sentry events.", e);
+    }
```

Comment:
I'd prefer to do this check only on `AndroidOptionsInitializer.initializeCacheDirs` maybe copying this piece of code in there, as the NDK needs this folder as well and it should be writable.
also at least for Android, if you `canWrite`, automatically you `canRead`, unless you root your device and change `chmod` manually.